src/main/java/org/codehaus/mojo/exec/ExecJavaMojo.java

```diff
@@ -358,6 +371,123 @@ public void execute() throws MojoExecutionException, MojoFailureException {
+                getLog().debug(
+                                "Can't inject " + runnableClass + "': " + re.getMessage() + ", will ignore injections",
+                                re);
+            }
+            final Constructor<? extends Runnable> declaredConstructor = runnableClass.getDeclaredConstructor();
+            declaredConstructor.setAccessible(true);
+            runnable = declaredConstructor.newInstance();
+        }
+        runnable.run();
+    }
+
+    private Object lookupParam(final java.lang.reflect.Parameter param) throws ComponentLookupException {
+        switch (param.getName()) {
+                // loose coupled to maven (wrapped with standard jvm types to not require it)
+            case "systemProperties": // Properties
+                return getSession().getSystemProperties();
+            case "systemPropertiesUpdater": // BiConsumer<String, String>
+                return propertiesUpdater(getSession().getSystemProperties());
+            case "userProperties": // Properties
+                return getSession().getUserProperties();
+            case "userPropertiesUpdater": // BiConsumer<String, String>
+                return propertiesUpdater(getSession().getUserProperties());
+            case "projectProperties": // Properties
+                return project.getProperties();
+            case "projectPropertiesUpdater": // BiConsumer<String, String>
+                return propertiesUpdater(project.getProperties());
+            case "highestVersionResolver": // Function<String, String>
+                return (Function<String, String>) ga -> {
+                    final int sep = ga.indexOf(':');
+                    if (sep < 0) {
+                        throw new IllegalArgumentException("Invalid groupId:artifactId argument: '" + ga + "'");
+                    }
+
+                    final org.eclipse.aether.artifact.Artifact artifact = new DefaultArtifact(ga + ":[0,)");
+                    final VersionRangeRequest rangeRequest = new VersionRangeRequest();
+                    rangeRequest.setArtifact(artifact);
+                    try {
+                        rangeRequest.setRepositories(project.getRemotePluginRepositories());
+                        final VersionRangeResult rangeResult = repositorySystem.resolveVersionRange(
+                                getSession().getRepositorySession(), rangeRequest);
+                        return String.valueOf(rangeResult.getHighestVersion());
+                    } catch (final VersionRangeResolutionException e) {
+                        throw new IllegalStateException(e);
+                    }
+                };
+                // standard bindings
+            case "session": // MavenSession
+                return getSession();
+            case "container": // PlexusContainer
+                return getSession().getContainer();
+            default: // Any
+                return getSession().getContainer().lookup(param.getType());
```

Comment:
`session.getContainer()` will be a `null` in Maven 4

Reply:
This mojo is mvn 3 so must be good, else not sure how i can get a generic lookup - will need a full rewrite for mvn 4 sadly so guess it is okish?

Reply:
injected plexus container for now to avoid to use the session getter which will be fine for now, when we'll move to maven 4 we'll use Lookup instead